onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc

```diff
@@ -913,7 +913,15 @@ PostLayoutTransformCostCheck(const api::GraphRef& graph, const api::NodeRef& nod
 Status TransformLayoutForEP(Graph& graph, bool& modified, const IExecutionProvider& execution_provider,
                             const DebugGraphFn& debug_graph_fn) {
   // sub graph recurse will be added later
-  auto api_graph = MakeApiGraph(graph, execution_provider.GetAllocator(OrtMemTypeDefault), nullptr);
+  auto cpu_allocator = execution_provider.GetAllocator(OrtMemTypeDefault);
```

Comment:
Would it be simpler for the caller to pass in the value returned by ExecutionProviders.GetDefaultCpuAllocator()?

Reply:
it's used in this way -
```
Status PartitionOrtFormatModel(onnxruntime::Graph& graph,
                               const ExecutionProviders& providers,
                               KernelRegistryManager& kernel_registry_manager,
                               SessionState& session_state) {

  layout_transformer::TransformLayoutFunction transform_layout_fn = layout_transformer::IsSupportedOpset(graph)
                                                                        ? layout_transformer::TransformLayoutForEP
                                                                        : nullptr;

  GraphPartitioner partitioner(kernel_registry_manager, providers);
  ORT_RETURN_IF_ERROR(partitioner.Partition(graph,
                                            session_state.GetMutableFuncMgr(),
                                            transform_layout_fn,
                                            GraphPartitioner::Mode::kOrtFormatLoad));

  return Status::OK();
}

```

So this is going to modify the function signature of `GraphPartitioner::Partition()` and some callback type definition, which is not a trivial one.

Reply:
In InferenceSessions.cs there is a lambda used so technically you'd only need to update TransformLayoutForEP and not the entire path through GraphPartitioner. 

https://github.com/microsoft/onnxruntime/blob/648bedf91a9e390fadbbabdbbb9d5239f5a9d201/onnxruntime/core/session/inference_session.cc#L944

Reply:
There are 2 places where `TransformLayoutForEP()` is called, and neither of them has a context with an existing allocator. If I make this change, I still need code to get allocator from the EP, and I need do this twice.

Reply:
Updated the code

Reply:
Is there an easier way to create lambdas in [session_state_test.cc](https://github.com/microsoft/onnxruntime/pull/15988/files#diff-2e498c0ec894ae05b0ddfb15f7bc4b927df6626c0497c47de85a5e808317d3f1)? is there a way to do currying in a simpler way